### PR TITLE
[PN-90] Add nodelets to create PointCloud2 messages from depth_registered and rgb image topics

### PR DESCRIPTION
**Pull request by jschornak-bdai**
## Description

- Replaces the previous `use_depth_registered_nodelets` boolean launch argument with the new `depth_registered_mode` string launch argument. This is converted into an enum within the `spot_driver.launch.py` launch file. This simplifies handling the three possible configurations of registered depth image publishing, and makes the configuration of these options more obvious to the user.
- Adds a new `publish_point_clouds` launch argument. When this argument is set to `true`, the driver launch file will add new nodelets that subscribe to the depth_registered and RGB image topics for each Spot camera and use these images to create and publish point clouds.
- Changes the executor type of the composable node container to `component_container_mt`. This is needed to allow the increased number of nodelets to work in parallel without blocking each other.
- Moves creation of the image processing nodelets to within `spot_driver.launch.py` instead of a separate launch file. This facilitates reuse of the same composable node container for different collections of nodelets.

## How to Test

Run `spot_driver.launch.py`, using the revised launch arguments to configure which nodes are brought up like in this example:
```
ros2 launch spot_driver spot_driver.launch.py publish_point_clouds:=true depth_registered_mode:=from_nodelets
```

Try all possible combinations of launch arguments to configure how the depth_registered and point cloud features are launched:

- `depth_registered_mode: disable` and `publish_point_clouds: false`: no `depth_registered` topics are published, and no nodelets are spawned.
- `depth_registered_mode: disable` and `publish_point_clouds: true`: same result as above -- `depth_registered_mode` overrides `publish_point_clouds` in this case.
- `depth_registered_mode: from_spot` and `publish_point_clouds: false`: No nodelets are spawned and no point cloud topics are published. Images are published to the `depth_registered` topics. 
- `depth_registered_mode: from_spot` and `publish_point_clouds: true`: One `depth_image_proc::PointCloudXyzrgbNode` nodelet is spawned for each camera. Both depth_registered and point cloud messages are published.
- `depth_registered_mode: from_nodelets` and `publish_point_clouds: false`: One RegisterNode nodelet is spawned for each camera. Images are published to the `depth_registered` topics. No point cloud topics are advertised.
- `depth_registered_mode: from_nodelets` and `publish_point_clouds: true`: One `depth_image_proc::RegisterNode` nodelet and one `depth_image_proc::PointCloudXyzrgbNode` nodelet is spawned for each camera. Both depth_registered and point cloud messages are published.

### CODE REVIEW

**Comment by mpickett-bdai** [comment]
@jschornak-bdai I attempted to run this on a spot and just wanted to confirm the use. I ran the standard `bdai spot driver <Robot Name>` but was unable to see the images the way I do on "main". The topics seem to show up, but no data is coming through. I also tried running just the spot_driver.launch.py on it's own and do not have see the images either. Maybe my docker is outdated. I'll try updating, but thought I'd mention what i've seen.

**Comment by mpickett-bdai** [comment]
Still not working with another docker.

**Review by jbarry-bdai — COMMENTED** [review]
I tested with the "from_nodelets" argument and that worked well.  However I did it through the bdai docker and didn't test all the options so this may require further testing by others to ensure the default behavior hasn't changed.  Sorry about that.

**Comment by mpickett-bdai** [comment]
@jbarry-bdai No worries. It's been a while since I've been running the spots so I thought it might be something on my end. I didn't try running with a specific argument (which might be it?) and just went for the default for `bdai spot driver` option as a check that that worked like it does on main. I also don't see a specific reason in the code for it to not work.

**Comment by jschornak-bdai** [comment]
> I didn't try running with a specific argument (which might be it?) and just went for the default for bdai spot driver option as a check that that worked like it does on main. I also don't see a specific reason in the code for it to not work.

If you're testing within the bdai Docker image, the version of the `ros_utilities` repo that is currently tagged as a dependency there is not compatible with versions of `spot_ros2` that are newer than 4f0483926e343da5af67f0bfc9252e64a4d9fabe. If you run the driver in the Docker container and you see a Python error about a failure to import the package, that's probably what's going wrong there.

**Comment by jbarry-bdai** [comment]
Oh right good point I had to manually update `ros_utilities` to the head of main for this to work inside bdai.  It's ok to merge this into spot_ros2 assuming the spot_ros2 submodule pointer is currently pointing at the head of `ros_utilities` but we can't merge the update to this hash into `bdai` until we've finished our move to the new ros utilities wrappers.

**Comment by mpickett-bdai** [comment]
I'll try again this morning. I wasnt seeing a specific error just empty images, but that could definitely be it.

**Comment by mpickett-bdai** [comment]
I'll try again this morning. I wasnt seeing a specific error just empty images, but that could definitely be it.

**Comment by mpickett-bdai** [comment]
Ran using PN-95 draft to make using it easy as @jschornak-bdai suggested and everything works well. I think the only thing to decide is how to keep in sync with ros_utilities and the bdai repo. Otherwise LGTM!

**Review by jbarry-bdai — APPROVED** [review]
This can be committed to spot ros2 but bdai should not be updated yet.

**Review by jbarry-bdai — APPROVED** [review]
This can be committed to spot ros2 but bdai should not be updated yet.

File: spot_driver/launch/spot_driver.launch.py
Comment (jschornak-bdai) [reply spot_driver/launch/spot_driver.launch.py]: I added an error message and early failure if an invalid option is provided. I agree that it's preferable to communicate the issue to the user rather than failing silently.

**Comment by amessing-bdai** [comment]
Can you update the example config file `<root>/spot_driver/config/spot_ros_example.yaml` with the new configs?

File: spot_driver/launch/spot_driver.launch.py
Comment (bhung-bdai) [inline spot_driver/launch/spot_driver.launch.py:103]: Is the default topic for the depth_image_proc being published on points? Why?

**Review by bhung-bdai — COMMENTED** [review]
I'm happy to give it an approval once Andrew's thing is addressed.

**Comment by jschornak-bdai** [comment]
> Can you update the example config file <root>/spot_driver/config/spot_ros_example.yaml with the new configs?

@amessing-bdai After a little investigation I realized that since these options are handled as launch arguments they cannot be configured through that config YAML file, since it's only defining values for parameters which are declared by the driver node. The way we'd configure these options is to change the command used to run the spot_ros2 driver launch files at the top level, for example like in [this PR](https://github.com/bdaiinstitute/bdai/pull/1045).

**Review by amessing-bdai — APPROVED** [review]

**Pull request merged**
